Convert accordions to accessible style

aydun opened the pull request:
Overview
----------------------------------------
Convert more accordions to `details`/`summary`

Before
----------------------------------------
Old style

After
----------------------------------------
New style

Technical Details
----------------------------------------
Some of these files have a `details` with a smarty conditional. In the old construct, accordions were open by default unless the `collapsed` class was included.  In the new, they are closed by default unless the `open` keyword is added. In some cases the presence of `collapsed` is determined by a smarty conditional.

This conditional needs to be preserved, but the logic reversed.  For ease of conversion, this PR does not attempt to negate the condition but uses the (inelegant but simple and existing) device of moving `open` to the other side of a smarty `{else}` - in some cases adding an `{else}`, removing it in others.  So `{if somecondition}collapsed{/if}` becomes `{if somecondition}{else}open{/if}` and `{if somecondition}{else}collapsed{/if}`becomes `{if somecondition}open{/if}`

This also changes the classes, removing `.crm-accordion-wrapper`, `.crm-accordion-header` and `.crm-master-accordion-header` and instead using `.crm-accordion-light` and `.crm-accordion-bold`  See [#29447#issuecomment-1959072462](https://github.com/civicrm/civicrm-core/pull/29447#issuecomment-1959072462) for details

Comments
----------------------------------------
~~Most of the non-conditional ones are in #29447  I'll rebase those out of here once those are merged. So focus on the last commit.~~
Change of plan: since the reviews are all on this PR, and the complications are in the js I have closed #29447 and we will use this as the consolidated PR.

aydun committed:
Remove line breaks in accordion divs

openvideotech commented on templates/CRM/Campaign/Form/Search/Common.tpl:
```diff
@@ -14,10 +14,10 @@
-  <div id="{$searchForm}" class="crm-accordion-wrapper crm-contribution_search_form-accordion {if $rows}collapsed{/if}">
-    <div class="crm-accordion-header {if !$votingTab} crm-master-accordion-header{/if}">
+  <details id="{$searchForm}" class="crm-accordion-wrapper crm-contribution_search_form-accordion" {if $rows}{else}open{/if}>
+    <summary class="crm-accordion-header {if !$votingTab} crm-master-accordion-header{/if}">
```
`.crm-master-accordion-header` is the utility class that changes the look of the entire accordion from dark bg/light text with border, to no header bg and no border. It only need to be applied to .crm-accordion-header for that to work (ie not the container div/details).

But since 5.69 there's been two utility classes to trigger these two UI versions, with a goal of removing `.crm-master-accordion-header`. In this instance it would be something like: `<details class="crm-accordion-{if !$votingTab}light{else}dark{/if}">.

vingle commented on templates/CRM/Campaign/Form/Search/Common.tpl:
```diff
@@ -14,10 +14,10 @@
-  <div id="{$searchForm}" class="crm-accordion-wrapper crm-contribution_search_form-accordion {if $rows}collapsed{/if}">
-    <div class="crm-accordion-header {if !$votingTab} crm-master-accordion-header{/if}">
+  <details id="{$searchForm}" class="crm-accordion-wrapper crm-contribution_search_form-accordion" {if $rows}{else}open{/if}>
+    <summary class="crm-accordion-header {if !$votingTab} crm-master-accordion-header{/if}">
```
`.crm-master-accordion-header` is the utility class that changes the look of the entire accordion from dark bg/light text with border, to no header bg and no border. It only need to be applied to .crm-accordion-header for that to work (ie not the container div/details).

But since 5.69 there's been two utility classes to trigger these two UI versions, with a goal of removing `.crm-master-accordion-header`. In this instance it would be something like: `<details class="crm-accordion-{if !$votingTab}light{else}dark{/if}">.

aydun replied:
Changes implemented

aydun committed:
Fix missing end div

aydun commented:
Updated based on @vingle's comments https://github.com/civicrm/civicrm-core/pull/29447#issuecomment-1959072462

vingle commented on templates/CRM/Contact/Form/Contact.tpl:
```diff
@@ -23,10 +23,10 @@
     {include file="CRM/common/formButtons.tpl" location="top"}
     </div>
 
-    <div class="crm-accordion-wrapper crm-contactDetails-accordion">
```
The JS from line 92 onwards interacts with the accordion class names here (called 'tabs' in the comments): e.g. hides them where there's no fields, opens if there's an error, and opens/closes all with the 'expand all tabs'/'collapse all tabs' link top right. This PR breaks that - so either the JS needs removing or rewriting (ref @colemanw). I guess this might depend on how far away the Form Builder version is.

herbdool replied:
I noticed the same issue. The JS could just target all details on the page, assuming there are no details that should *not* be toggled.

herbdool replied:
@vingle this is what I changed in this file to get the expand/collapse all tabs to work:

```
$('a#expand').click( function() {
      if( $(this).attr('href') == '#expand') {
        var message = {/literal}"{ts escape='js'}Collapse all tabs{/ts}"{literal};
        $(this).attr('href', '#collapse');
        $('.crm-form-block details:not([open])').crmAccordionToggle();
      }
      else {
        var message = {/literal}"{ts escape='js'}Expand all tabs{/ts}"{literal};
        $('.crm-form-block details[open]').crmAccordionToggle();
        $(this).attr('href', '#expand');
      }
      $(this).html(message);
      return false;
    });
```

vingle replied:
That looks good to me. I guess at some point we could add details about crmAccordionToggle to the docs.

herbdool replied:
@aydun do you prefer I make a PR against your branch, or is including a code snippet enough?

aydun replied:
@herbdool - probably best if you do a PR against mine - thanks.

herbdool replied:
@aydun here's the PR https://github.com/aydun/civicrm-core/pull/1

vingle commented:
To make matters more complex, some (one so far) of the template files here don't contain JS but are nested inside other files that do contain inline JS that references the nested template – but aren't in this PR as they've not changed.

E.g. https://github.com/civicrm/civicrm-core/blob/2177f1c1bd81b403a1cc23d72a416c230b557304/templates/CRM/Activity/Form/Activity.tpl#L205 in `crm/activity/form/activity.tpl` is another script for opening the accordion if there's an error. But there's no accordion on that template, so it's not in this PR, the accordions are in related files, e.g. `CRM/Activity/Form/FollowUp.tpl`

Again if this on a form that's soon to be FormBuildered into something soon, then maybe it's an acceptable short-term regression. But the existence of this pattern suggests not all the JS consequences of these changes can be tested easily.

vingle commented on templates/CRM/Case/Form/ActivityTab.tpl:
```diff
@@ -10,10 +10,10 @@
+    </summary>
 
     <div id="activities" class="crm-accordion-body padded">
     <div class="crm-collapsible crm-search_filters-accordion collapsed">
```
There's another accordion here (pattern 3), light bg. Worth swapping for `<details class="crm-accordion-light">… etc`

vingle commented on templates/CRM/Contribute/Form/AdditionalPayment.tpl:
```diff
@@ -103,7 +103,7 @@
       </div>
```
This closing div closes the crm-accordion-body too soon so the check field is outside the border. This `</div>` should be removed, and the one below (line 106) brought back.

<img width="533" alt="image" src="https://github.com/civicrm/civicrm-core/assets/1175967/ec3c598b-93e9-4cac-ba29-0749a2b8bf8f">

shaneonabike replied:
I saw this one also but I think this should be a separate PR later since it existed before we did these changes.

I created a new ticket to address this at [issue#69](https://lab.civicrm.org/dev/user-interface/-/issues/69)

herbdool commented:
test this please

vingle commented:
I'm reviewing this WIP rather than the shorter PR. Breaking down comments on the 64 files:

### Possibly deprecated? (4)

These screens appeared to have SK/FB versions that ship now with core so I can't test, but maybe aren't needed:

- Campaign/Form/Search/Common.tpl
- Campaign/Form/Survey/Results.tpl
- Campaign/Form/Task/Reserve.tpl
- Contact/Form/Search/Builder.tpl

### Cannot find (5)

The code looks good with all of these, but I can't visually check (maybe someone else finds them). None of them have linked JS issue so may be ok to merge.

- Custom/Form/Search.tpl
- Mailing/Form/Approve.tpl
- Member/Form/Membership.tpl
- Report/Form/Tabs/FieldSelection.tpl
- Report/Form/Tabs/Filters.tpl

### Working but do these need to be accordions? (5)

Recently [an accordion was removed](https://github.com/civicrm/civicrm-core/pull/29446) as adding much to the UI/UX, and the same could be said for these. The last one (FinancialAccount) would definitely be better *not* an accordion as it's creating an empty accordion when not populated.

- Case/Form/CaseFilter.tpl - civicrm/case
- Contribute/Form/ContributionPage/Premium.tpl - civicrm/admin/contribute/custom?action=update&reset=1&id=1&selectedChild=premium
- SMS/Form/Group.tpl - civicrm/sms/send?reset=1 (install & config SMS dummy ext)
- SMS/Form/Schedule.tpl - civicrm/sms/send?reset=1
- Financial/Form/FinancialAccount.tpl - civicrm/admin/financial/financialAccount?reset=1 + add

### Remove from PR (1)

This template has two types of accordion, and this PR fixes one of them. https://github.com/civicrm/civicrm-core/pull/29533 fixes both

- Case/Form/ActivityTab.tpl - civicrm/contact/view/case

### Broken - Ajax issue (3)

Should be fixable by correcting the Ajax function on each

- Contact/Page/View/GroupContact.tpl - civicrm/contact/view -> Groups
- Contribute/Form/Contribution.tpl - civicrm/contribute/add
- Pledge/Form/Pledge.tpl - civicrm/pledge/add?reset=1&action=add&context=standalone

### Broken - possibly was already (1)

- common/fatal.tpl - make an error, ie civicrm/payment

Hard to test the fatal error accordion as I'm not sure a) if that page is behaving as intended before this PR; and b) don't know the conditions to display the accordion. Commenting out the conditionals shows a working accordion, but there's nothing inside it, but not sure if that's the fault of this PR or not.

### Good (29)

- Activity/Form/Search.tpl
- Case/Form/CaseView.tpl
- Case/Form/Search.tpl
- Case/Page/CustomDataView.tpl
- Contact/Form/Search/BasicCriteria.tpl
- Contact/Form/Task/EmailCommon.tpl
- Contact/Form/Task/PDFLetterCommon.tpl
- Contact/Form/Task/SMSCommon.tpl
- Contact/Page/DedupeFind.tpl
- Contribute/Form/ContributionPage/Widget.tpl
- Contribute/Form/ContributionView.tpl
- Contribute/Form/Search.tpl
- Contribute/Form/Search/ContributionRecur.tpl
- Contribute/Form/Task/PDFLetter.tpl
- Contribute/Page/ContributionRecurPayments.tpl
- Custom/Form/CustomData.tpl
- Event/Form/ManageEvent/Repeat.tpl
- Event/Form/Search.tpl
- Event/Form/SearchEvent.tpl
- Financial/Form/Search.tpl
- Group/Form/Search.tpl
- Member/Form/MembershipView.tpl
- Member/Form/Search.tpl
- Profile/Form/Search.tpl
- Report/Page/InstanceList.tpl
- Report/Page/TemplateList.tpl
- UF/Form/AdvanceSetting.tpl
- Activity/Selector/Selector.tpl
- Admin/Form/MessageTemplates.tpl

### Functional, but issues (16)

These accordions work as expected and are more accessible, but some javascript isn't working as intended. Will describe the JS issues in next comment.

Ideally the JS would be fixed, but if they can't, it could be argued that accessibility is more important than the broken JS, especially for screens soon to be replaced by SK/FB:

- Activity/Form/FollowUp.tpl - civicrm/activity?reset=1&action=add&context=standalone
- Case/Form/Activity.tpl - civicrm/contact/view/case (add activity modal)
- Contact/Form/Contact.tpl - civicrm/contact/add
- Contact/Form/Edit/Address.tpl - civicrm/contact/add
- Contact/Form/Edit/CommunicationPreferences.tpl - civicrm/contact/add
- Contact/Form/Edit/CustomData.tpl - civicrm/contact/add
- Contact/Form/Edit/Demographics.tpl - civicrm/contact/add
- Contact/Form/Edit/Notes.tpl - civicrm/contact/add
- Contact/Form/Edit/TagsAndGroups.tpl - civicrm/contact/add
- Contact/Import/Form/Preview.tpl - civicrm/import/contact > page 3
- Contribute/Form/AdditionalPayment.tpl - civicrm/payment?action=add&reset=1&id=1
- Core/Form/RecurringEntity.tpl - civicrm/activity?reset=1&action=add&context=standalone
- Custom/Page/CustomDataView.tpl - viewing anything saved from a custom field
- Financial/Form/BatchTransaction.tpl - civicrm/batchtransaction?reset=1&bid=1 (after creating a batch)
- Form/attachment.tpl - civicrm/activity?reset=1&action=add&context=standalone, civicrm/activity/email/add
- Pledge/Form/Search.tpl - civicrm/pledge/search?reset=1

aydun commented:
> I'm reviewing this WIP rather than the shorter PR. Breaking down comments on the 64 files:

That's a huge review @vingle !

vingle commented:
It just kept expanding @aydun, like an infinite accordion……

vingle commented:
@herbdool – do you know which file is holding up that PR?

I'm fine with `crm-toggle` and then target `details.crm-toggle`? Or `.crm-accordion-toggle` if it's simpler.

While testing this I realised there were other toggle functions being used with an accordion-style icon - ie that aren't accordions but have an accordion appearance to toggle another region's visibility(!). But they seem to be accessible so aren't a worry for now. Eg.

![image](https://github.com/civicrm/civicrm-core/assets/1175967/3eacfbb2-2467-45f0-a706-d15238b277a2)

aydun committed:
Convert accordions to `<details>`/`<summary>`

aydun commented:
I have rebased and excluded `templates/CRM/Case/Form/ActivityTab.tpl` which is now in #29533

aydun commented:
@herbdool thanks - rebased again to include #29449

herbdool committed:
Fix expand/collapse all tabs
Fix open accordions on error
Fix highlightTabs()
Fix crm-ajax-accordion
Clean up fatal error accordion

herbdool commented:
@aydun I made a PR against your branch: https://github.com/aydun/civicrm-core/pull/1

aydun commented:
> @aydun I made a PR against your branch: [aydun#1](https://github.com/aydun/civicrm-core/pull/1)

Thanks! - merged

herbdool commented:
@aydun another PR against yours: https://github.com/aydun/civicrm-core/pull/2. I've tested and this works.

herbdool commented:
A fix to highlighting tabs: https://github.com/aydun/civicrm-core/pull/3

highfalutin committed:
remove code that will never run from contact import form

herbdool committed:
update crmAccordionToggle()
Update templates/CRM/Activity/Form/Activity.tpl
Update templates/CRM/Campaign/Form/Gotv.tpl
Update templates/CRM/Campaign/Form/Task/Reserve.tpl
Update templates/CRM/Contact/Form/Contact.tpl
Update templates/CRM/Contact/Form/Contact.tpl
Update templates/CRM/Contact/Form/Contact.tpl
Update templates/CRM/Contact/Form/Search/AdvancedCriteria.tpl
Update templates/CRM/Financial/Form/BatchTransaction.tpl
Apply suggestions from code review
Fix Pledge form ajax

Co-authored-by: highfalutin <highfalutin@users.noreply.github.com>

herbdool commented:
Another PR https://github.com/aydun/civicrm-core/pull/5. Also based on @highfalutin's suggestion on removing call to `crmAccordionToggle()` where we can.

herbdool commented:
@aydun you may want to squash merge the PR https://github.com/aydun/civicrm-core/pull/5 as I made a few commits to fix different suggestions from @highfalutin

vingle committed:
Removes outdated accordion classes in wrapper

Cannot see function of them - but it creates some UI quirks.

shaneonabike commented:
test this please

shaneonabike commented on templates/CRM/Contribute/Form/Contribution.tpl:
```diff
@@ -405,12 +405,12 @@
       <div class="accordion ui-accordion ui-widget ui-helper-reset">
```
Personally I think that this should just be removed. We don't use jquery accordion and with new method it handles itself in the browser.

aydun commented:
I've rebased on master and squashed various commits - hopefully without losing and changes or attributions!

herbdool commented:
In the build I see this error: 

> Patch failed at 0001 Fixes Accordion Help button CSS, following PRs #29551 

It says there are no conflicts so I'm not sure why that patch doesn't apply.

colemanw commented:
@civicrm-builder retest this please

aydun committed:
Remove unneeded comments: <!-- /.crm-accordion-body -->

highfalutin commented:
@aydun I think it's time to remove the "WIP" from this.

highfalutin commented:
@civicrm-builder retest this please

highfalutin commented:
@aydun, looks like this needs one final rebase.

aydun commented:
@highfalutin Ok, rebased

mattwire commented:
Merging based on discussion with @highfalutin and @aydun

aydun commented:
Thank you sprinters for getting this one over the line!
